### ISSUE TITLE
Apply configurable timeouts when sending emails

### ISSUE THREAD

**Issue by amanteaux** [bbottema/simple-java-mail#85]
I will start with my use case: I encountered an OutOfMemoryError on a production server that was due to emails that were queuing.
To give an insight of what where going, here is what was contained in the heap space:
![image](https://user-images.githubusercontent.com/3846854/28031753-a2d1ef36-65a8-11e7-8d4b-57bbfd42ffeb.png)

When the production team detected the error, they restarted the server, here is an extract of the logs:
[logs_exact.txt](https://github.com/bbottema/simple-java-mail/files/1136548/logs_exact.txt)

I think the problem is that in `MailSender` the executor threads were all waiting to get a connection from the SMTP server and during this time mails were added to the executor queue until the server run out of memory. So the real problem is that the SMTP server was down and no timeout are configured in `SMTPTransport` class.
To solve this problem, a solution would be to add a timeout in  `MailSender` that will cancel the sending task if it takes more that X seconds/minutes like in https://stackoverflow.com/a/2759040/3790208.

Moreover, though the code is really nicely written in `MailSender`, I noticed 2 things that can be improved:
- line 165: a `Runnable` should be provided instead of a `Thread` object that adds a little overhead,
- line 159: `sendMailClosure` should be wrapped in a try catch to at least show the date of the exception.

I can make a pull request if that is ok.

**Comment by bbottema** [comment]
By all means, give it a shot. I'm on my phone right now, I'll give a proper reply when I get a chance.

**Comment by amanteaux** [comment]
Thank you for your quick answer! I will try to put up something this week.

**Comment by bbottema** [comment]
Released in 4.3.0.

**Comment by bbottema** [comment]
Current 9.x usage (verified against 9.3.2):

```java
Mailer mailer = MailerBuilder
    .withSMTPServer("smtp.example.com", 587)
    .withSessionTimeout(10_000) // milliseconds
    .buildMailer();
```

The old mutable setter moved to the mailer builder in 5.0. This single value applies to the SMTP connect, read, and write timeouts. See the [5.0 builder migration](https://www.simplejavamail.org/migration-notes-5.0.0.html#section-what-changed) and [current programmatic configuration](https://www.simplejavamail.org/configuration.html#section-programmatic-api-other).
